### PR TITLE
Port `ArrayPosition` and `ArrayPositions` to `functions-array` subcrate

### LINKED ISSUE #9616: Port `ArrayPosition` and `ArrayPositions` to `functions-array` subcrate
### Is your feature request related to a problem or challenge?

`ArrayPosition` and `ArrayPositions` functions need to be ported to new `functions-array` subcreate in terms of https://github.com/apache/arrow-datafusion/issues/9285.

### Describe the solution you'd like

Port `ArrayPosition` and `ArrayPositions` to `functions-array` subcrate

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by erenavsarogullari**
## Which issue does this PR close?
Closes #9616.

## What changes are included in this PR?
This PR aims to do following changes in terms of Epic https://github.com/apache/arrow-datafusion/issues/9285:
`ArrayPosition` and `ArrayPositions` are ported to `functions-array` subcrate.

## Are these changes tested?
Yes, all `array.slt` based `ArrayPosition` and `ArrayPositions` tests are passed.

## Are there any user-facing changes?
No

### CODE REVIEW

**Comment by erenavsarogullari** [comment]
@jayzhan211 Created this PR in terms of our previous discussion on the `kernels/udf` structure. Please let me know if you have any concern. Also, will be working on current test failures. Just fyi.

File: datafusion/functions-array/src/position.rs [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+#[derive(Debug)]
+pub(super) struct ArrayPosition {
+    signature: Signature,
+    aliases: Vec<String>,
+}
+impl ArrayPosition {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::variadic_any(Volatility::Immutable),
```
Comment (jayzhan211) [inline datafusion/functions-array/src/position.rs:56]: I think the test should be fixed if you change the signature back to the previous one `array_and_element_and_optional_index`

Comment (erenavsarogullari) [reply datafusion/functions-array/src/position.rs]: Thanks for quick feedback, it is fixed.

**Comment by jayzhan211** [comment]
Remind to add the roundtrip test

**Review by Weijun-H — APPROVED** [review]
Nice work! Thanks @erenavsarogullari

**Review by alamb — APPROVED** [review]
Thank you @erenavsarogullari 

I took the liberty of merging up from main to fix the merge conflicts and I removed the dependency on arrow-ord to fix the datafusion-cli CI check

Thank you @Weijun-H  for the review

File: datafusion/functions-array/src/position.rs [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+            }
+        } else {
+            // We will get null if we got null in the array, so we don't need to check
+        }
+    }
+
+    generic_position::<O>(list_array, element_array, arr_from)
+}
+
+fn check_datatypes(name: &str, args: &[&ArrayRef]) -> datafusion_common::Result<()> {
```
Comment (jayzhan211) [inline datafusion/functions-array/src/position.rs:144]: We can use function in crate::utils

File: datafusion/functions-array/src/position.rs [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    fn name(&self) -> &str {
+        "array_position"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> datafusion_common::Result<DataType> {
+        Ok(match arg_types[0] {
```
Comment (jayzhan211) [inline datafusion/functions-array/src/position.rs:82]: I don't think the type checking is necessary in `return_type` 🤔

File: datafusion/functions-array/src/position.rs (hunk 2 above)
Comment (erenavsarogullari) [reply datafusion/functions-array/src/position.rs]: Addressed

File: datafusion/functions-array/src/position.rs (hunk 3 above)
Comment (erenavsarogullari) [reply datafusion/functions-array/src/position.rs]: Addressed by reflecting both `array_position` and `array_positions`.

**Comment by erenavsarogullari** [comment]
Thanks everyone for the reviews and feedbacks.

**Comment by jayzhan211** [comment]
Thanks @erenavsarogullari and @alamb , @Weijun-H for the review

**Pull request merged**